packages/asset/contracts/AssetReveal.sol

```diff
@@ -0,0 +1,266 @@
+            tokenId,
+            data.creator,
+            data.tier,
+            data.creatorNonce,
+            amount
+        );
+    }
+
+    /// @notice Burn multiple assets to be able to reveal them later
+    /// @dev Can be used to burn multiple copies of the same token id, each copy will be revealed separately
+    /// @param tokenIds the tokenIds of the assets to burn
+    /// @param amounts the amounts of the assets to burn
+    function revealBatchBurn(
+        uint256[] calldata tokenIds,
+        uint256[] calldata amounts
+    ) external {
+        require(tokenIds.length == amounts.length, "Invalid input");
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            revealBurn(tokenIds[i], amounts[i]);
+        }
+    }
+
+    /// @notice Reveal assets to view their abilities and enhancements
+    /// @dev Can be used to reveal multiple copies of the same token id
+    /// @param signature Signature created on the TSB backend containing REVEAL_TYPEHASH and associated data, must be signed by authorized signer
+    /// @param creator The original creator of the assets
+    /// @param prevTokenId The tokenId of the unrevealed asset
+    /// @param amount The amount of assets to reveal (must be equal to the length of revealHashes)
+    /// @param metadataHashes The array of hashes for asset metadata
+    /// @param recipient The recipient of the revealed assets
+    function revealMint(
+        bytes memory signature,
+        address creator,
+        uint256 prevTokenId,
+        uint256 amount,
+        string[] memory metadataHashes,
+        address recipient
+    ) public {
+        require(
+            authValidator.verify(
+                signature,
+                _hashReveal(creator, prevTokenId, amount, metadataHashes)
+            ),
+            "Invalid signature"
+        );
+
+        IAsset.AssetData memory data = prevTokenId.getData();
+        require(!data.revealed, "Asset: already revealed");
+        require(
+            data.creator == creator,
+            "Asset: creator does not match prevTokenId"
+        );
+        require(amount == metadataHashes.length, "Invalid amount");
+
+        uint256[] memory tokenIdArray = new uint256[](amount);
+        uint256[] memory tokenCountArray = new uint256[](amount);
+        uint256 uniqueTokenCount = 0;
+
+        // for each asset, set the data
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 tokenId;
+            if (
+                assetContract.getTokenIdByMetadataHash(metadataHashes[i]) != 0
+            ) {
+                tokenId = assetContract.getTokenIdByMetadataHash(
+                    metadataHashes[i]
+                );
+            } else {
+                uint16 revealNonce = ++revealNonces[data.creator][prevTokenId];
+
+                tokenId = TokenIdUtils.generateTokenId(
+                    data.creator,
+                    data.tier,
+                    data.creatorNonce,
```

Comment:
`data.creatorNonce` data comes from `IAsset.AssetData memory data = prevTokenId.getData();` which returns the nonce in the tokenId, not the last one.
something fishy here

Reply:
Nothing fishy but I may have not explained it sufficiently.
Here is some more information.

When creator mints a new asset it is assigned a creator nonce so that all new tokens even with same statistics end up having different tokenId.

When an asset is revealed we want to keep the old creatorNonce as the creator is not in fact making a new token, we are just revealing it's abilities and enhancements. In this case the differentiator that will make sure all different revealed copies get unique ids based on their abilities and enhancements is the revealNonce.

Hope that makes sense and if you need some more info or you have ideas in this matter let me know.